Rasa CLI plugins (#6447)
RasaHQ/rasa: docs/docs/command-line-interface.mdx

```text [rasa x --help]
```

## Command Line Plugins

You can extend the Rasa Open Source command line interface with custom plugins.
Plugins extending the `rasa` command line interface can implement new or custom
behavior which is not included within Rasa Open Source by default.

:::note
It is currently not possible to overwrite existing `rasa` commands. For example it is
not possible to use a plugin `rasa-train` to override the `rasa train` command.
:::

### Installing a Plugin

A command line plugin is nothing more than an executable file, whose name begins with
`rasa-`. Once it's installed, it is available on the command line via
`rasa <plugin name>`.

> To install a command line plugin, move its executable somewhere on your `PATH`.

Reviewer: i'd tell the user either how to view their path, or how to add the directory with the executable to their path

Reviewer: maybe by moving the

To make the plugin discoverable, it has to be in your `PATH` environment. You can see
all potential directories in your `PATH` by running `echo $PATH`.
Assuming that `/usr/local/bin` is in your `PATH`, make your plugin discoverable by
moving it to there:

```bash
sudo mv ./rasa-hello  /usr/local/bin
```

Stuff up here and linking from the other one like "install your plugin to make it discoverable". this would make sure users know how to install whether writing their own or using one someone else wrote

Make sure that the executable name begins with `rasa-`.

> ### Listing Available Plugins

Reviewer 2: The part that is not clear to me. How can I take advantage of plugins created by other people? Where can find them?

Reviewer 3: I agree and I think it relates a bit to my comment:
> what could be nice is to have a pypi package (or other) that is named rasa-hello and is an example of a plugin. That way you could reference it in the docs + it would give a boilerplate for user (they could fork it and play with it). But maybe not at this stage 🤷

if we have a lightweight setup here:
- setup a `rasa-plugins` public monorepo
- create a `rasa-hello` plugin there
- welcome pull requests to the repo
- have a quick-and-dirty way to install the plugins (basically clone the repo + manipulate the path)

I think that could work easily and help with discovery / contribution. WDYT @TyDunn @wochinge ?

Author: @koaning How about using `rasa-lit` for this?

Author: > How can I take advantage of plugins created by other people?

I think that's something they would need to document in their README as this might be different for each programming language. 

> Where can find them?

We can start listing some here as they evolve. A good start would be `rasa lit` in my opinion. 

> setup a rasa-plugins public monorepo

I don't think that one repository for all plugins makes sense. Take `rasa-lit` as example: It stands for itself and the fact that it can be used as CLI plugin for Rasa Open Source is only a nice to have for it. 

> have a quick-and-dirty way to install the plugins (basically clone the repo + manipulate the path)

That would be a simply `pip install` btw.

I can obviously create a quick `rasa-hello` repo, but I don't think it's necessary because
1) `rasa-lit` can serve as example
2) The target group for this is developers who are developing some tool in the rasa ecosystem. They don't need any boilerplate to start out, they just need to expose their existing project as executable so that Rasa Open Source can find it. The way this is done is different per build tool / programming language, but is usually very well documented.

You can use `rasa plugins` to list command line plugins which are discoverable by
Rasa Open Source. Rasa Open Source discovers command line plugins by searching your
`PATH` environment for executables which begin with the prefix `rasa-`.

### Writing a Plugin

You can write a plugin in any programming language or script that allows you

> to write command-line commands. The command which your plugin is providing, is
> determined based on the name of our plugin's executable. Everything after the first dash

Reviewer: ```suggestion
to write command-line commands. The command that your plugin provides is
determined based on the name of your plugin's executable. Everything after the first dash
```
